src/transformState.js

```diff
@@ -6,6 +6,32 @@
  */
 function transformState(state, actions) {
   // write code here
+  for (let n = 0; n < actions.length; n++) {
+    switch (actions[n].type) {
+      case 'addProperties':
+        const actionEntries = Object.entries(actions[n].extraData);
+
+        for (const [key, value] of actionEntries) {
+          state[key] = value;
+        }
+
+        break;
+
+      case 'removeProperties':
+        for (let i = 0; i < actions[n].keysToRemove.length; i++) {
+          delete state[actions[n].keysToRemove[i]];
+        }
+
+        break;
+
+      case 'clear':
+        for (let i = (Object.keys(state).length - 1); i >= 0; i--) {
+          delete state[Object.keys(state)[i]];
+        }
```

Comment:
regarding your comments - you don't need a variable because you have to modify this `state` anyway
You don't need to use Object.keys just to get the length of the state, you can just use state.length or for...in loop. Alternatively you can probably just do state = {} to clear the state. Try some of it but don't use Object.keys here and don't add redundant variables

Reply:
Cannot use state = {} as state in this assignment is a reference passed as a parameter rather than a clone
And with state.length... back when I tried doing that on the Mate website, it outputted errors, that's why I've avoided doing that, I will try a for...in loop